src/model.py

```diff
@@ -496,21 +513,58 @@ def _create_response(
                 )
             )
 
+        # logprobs
+        if additional_outputs["return_logprobs"]:
+            if "prev_lens_logprobs" not in request_output_state:
+                request_output_state["prev_lens_logprobs"] = [0] * len(
+                    request_output.outputs
+                )
+            logprobs = []
+            for i in range(len(request_output.outputs)):
+                output = request_output.outputs[i]
+                if output.logprobs is None:
+                    logprobs.append("null".encode("utf-8"))
+                    continue
+                prev_len = request_output_state["prev_lens_logprobs"][i]
+                request_output_state["prev_lens_logprobs"][i] = len(output.logprobs)
+                logprobs_py = []
+                for logprob_d_vllm in output.logprobs[prev_len:]:
+                    logprob_d_py = {}
+                    for token_id, logprob_vllm in logprob_d_vllm.items():
+                        logprob_d_py[token_id] = {
+                            "logprob": logprob_vllm.logprob,
+                            "rank": logprob_vllm.rank,
+                            "decoded_token": logprob_vllm.decoded_token,
+                        }
```

Comment:
What's the reference for this implementation? How did you decide to return rank, decoded_token, etc. and how many top N logprobs to return?

Reply:
> What's the reference for this implementation?
https://github.com/vllm-project/vllm/blob/main/vllm/sequence.py#L53-L54

> How did you decide to return rank, decoded_token, etc
They are vLLM specific from https://github.com/vllm-project/vllm/blob/main/vllm/sequence.py#L33-L47

> how many top N logprobs to return?
This is controlled by the user by setting the input parameter `logprobs`. vLLM returns up to logprobs + 1. https://github.com/vllm-project/vllm/blob/main/vllm/sampling_params.py#L143-L149